File: libraries/apollo-runtime-java/src/main/java/com/apollographql/apollo/runtime/java/network/ws/protocol/GraphQLWsProtocol.java
package com.apollographql.apollo.runtime.java.network.ws.protocol;

import com.apollographql.apollo.api.ApolloRequest;
import com.apollographql.apollo.api.ImmutableMapBuilder;
import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.http.DefaultHttpRequestComposer;
import com.apollographql.apollo.runtime.java.network.ws.WebSocketConnection;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.function.Supplier;

/**
 * An {@link WsProtocol} for <a href="https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md">the GraphQL over WebSocket
 * Protocol</a>. It can carry queries in addition to subscriptions over the websocket.
 */
public class GraphQLWsProtocol extends WsProtocol {
  private Supplier<Map<String, Object>> connectionPayload;
  private WsFrameType frameType;
  private Map<String, Object> pingPayload;
  private Map<String, Object> pongPayload;
  private long connectionAcknowledgeTimeoutMs;
  private long pingIntervalMillis;

  private ScheduledExecutorService pingExecutor = Executors.newSingleThreadScheduledExecutor();

  public GraphQLWsProtocol(
      WebSocketConnection webSocketConnection,
      Listener listener,
      Supplier<Map<String, Object>> connectionPayload,
      WsFrameType frameType,
      Map<String, Object> pingPayload,
      Map<String, Object> pongPayload,
      long connectionAcknowledgeTimeoutMs,
      long pingIntervalMillis
  ) {
    super(webSocketConnection, listener);
    this.connectionPayload = connectionPayload;
    this.frameType = frameType;
    this.pingPayload = pingPayload;
    this.pongPayload = pongPayload;
    this.connectionAcknowledgeTimeoutMs = connectionAcknowledgeTimeoutMs;
    this.pingIntervalMillis = pingIntervalMillis;
  }

  @Override
  public void connectionInit() {
    Map<String, Object> message = new ImmutableMapBuilder<String, Object>().put("type", "connection_init").build();
    Map<String, Object> payload = connectionPayload.get();
    if (payload != null) {
      message.put("payload", payload);
    }
    sendMessageMap(message, frameType);

    Map<String, Object> map = receiveMessageMap(connectionAcknowledgeTimeoutMs);
    if (map == null) {
      listener.networkError(new IOException("Connection closed or timeout while waiting for connection_ack"));
    } else {
      Object type = map.get("type");
      if ("connection_error".equals(type)) {
        listener.networkError(new IOException("connection_error received " + map));
      } else if (!"connection_ack".equals(type)) {
        System.out.println("unknown graphql-transport-ws message while waiting for connection_ack: " + type);
      }
    }
  }

  @SuppressWarnings("unchecked")
  @Override
  public void handleServerMessage(Map<String, Object> messageMap) {
    String type = (String) messageMap.get("type");
    switch (type) {
      case "next":
        listener.operationResponse((String) messageMap.get("id"), (Map<String, Object>) messageMap.get("payload"));
        break;
      case "error":
        listener.operationError((String) messageMap.get("id"), (Map<String, Object>) messageMap.get("payload"));
        break;
      case "complete":
        listener.operationComplete((String) messageMap.get("id"));
        break;
      case "ping":
        Map<String, Object> map = new ImmutableMapBuilder<String, Object>().put("type", "pong").build();
        if (pongPayload != null) {
          map.put("payload", pongPayload);
        }
        sendMessageMap(map, frameType);
        break;
      case "pong":
        // Nothing to do, the server acknowledged one of our pings
        break;
      default:
        // Unknown message
        break;
    }
  }

  @Override
  public <D extends Operation.Data> void startOperation(ApolloRequest<D> request) {
    sendMessageMap(
        new ImmutableMapBuilder<String, Object>()
            .put("type", "subscribe")
            .put("id", request.getRequestUuid().toString())
            .put("payload", DefaultHttpRequestComposer.Companion.composePayload(request))
            .build(),
        frameType
    );
  }

  @Override
  public <D extends Operation.Data> void stopOperation(ApolloRequest<D> request) {
    sendMessageMap(
        new ImmutableMapBuilder<String, Object>()
            .put("type", "complete")
            .put("id", request.getRequestUuid().toString())
            .build(),
        frameType
    );
  }

  @Override
  public void run() {
    if (pingIntervalMillis > 0) {
      pingExecutor.scheduleWithFixedDelay(() -> {
        Map<String, Object> map = new ImmutableMapBuilder<String, Object>()
            .put("type", "ping")
            .build();
        if (pingPayload != null) {
          map.put("payload", pingPayload);
        }
        sendMessageMap(map, frameType);
      }, pingIntervalMillis, pingIntervalMillis, TimeUnit.MILLISECONDS);
    }

    super.run();
  }

  public static class Factory implements WsProtocol.Factory {
    private Supplier<Map<String, Object>> connectionPayload;
    private WsFrameType frameType;
    private Map<String, Object> pingPayload;
    private Map<String, Object> pongPayload;
    private long connectionAcknowledgeTimeoutMs;
    private long pingIntervalMillis;

    public Factory() {
      this(() -> null, WsFrameType.Text, null, null, 60000, -1);
    }

    public Factory(
        Supplier<Map<String, Object>> connectionPayload,
        WsFrameType frameType,
        Map<String, Object> pingPayload,
        Map<String, Object> pongPayload,
        long connectionAcknowledgeTimeoutMs,
        long pingIntervalMillis
    ) {
      this.connectionPayload = connectionPayload;
      this.frameType = frameType;
      this.pingPayload = pingPayload;
      this.pongPayload = pongPayload;
      this.connectionAcknowledgeTimeoutMs = connectionAcknowledgeTimeoutMs;
      this.pingIntervalMillis = pingIntervalMillis;
    }

    @Override public String getName() {
      return "graphql-transport-ws";
    }

    @Override public WsProtocol create(
        WebSocketConnection webSocketConnection,
        Listener listener
    ) {
      return new GraphQLWsProtocol(
          webSocketConnection,
          listener,
          connectionPayload,
          frameType,
          pingPayload,
          pongPayload,
          connectionAcknowledgeTimeoutMs,
          pingIntervalMillis
      );
    }
  }

}

File: libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/GraphQLWsProtocol.kt
package com.apollographql.apollo3.network.ws

import com.apollographql.apollo3.annotations.ApolloDeprecatedSince
import com.apollographql.apollo3.annotations.ApolloDeprecatedSince.Version.v3_2_3
import com.apollographql.apollo3.api.ApolloRequest
import com.apollographql.apollo3.api.Operation
import com.apollographql.apollo3.api.http.DefaultHttpRequestComposer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout


/**
 * An [WsProtocol] that uses https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md
 * It can carry queries in addition to subscriptions over the websocket
 */
class GraphQLWsProtocol(
    private val connectionPayload: suspend () -> Map<String, Any?>? = { null },
    private val pingPayload: Map<String, Any?>? = null,
    private val pongPayload: Map<String, Any?>? = null,
    private val connectionAcknowledgeTimeoutMs: Long,
    private val pingIntervalMillis: Long,
    private val frameType: WsFrameType,
    webSocketConnection: WebSocketConnection,
    listener: Listener,
    private val scope: CoroutineScope,
) : WsProtocol(webSocketConnection, listener) {

  @Deprecated("Use the constructor with connectionPayload as a lambda instead",
      ReplaceWith("Factory({ connectionPayload }, pingIntervalMillis, pingPayload, pongPayload, connectionAcknowledgeTimeoutMs)"))
  @ApolloDeprecatedSince(v3_2_3)
  constructor(
      connectionPayload: Map<String, Any?>? = null,
      pingPayload: Map<String, Any?>? = null,
      pongPayload: Map<String, Any?>? = null,
      connectionAcknowledgeTimeoutMs: Long,
      pingIntervalMillis: Long,
      frameType: WsFrameType,
      webSocketConnection: WebSocketConnection,
      listener: Listener,
      scope: CoroutineScope,
  ) : this(
      connectionPayload = { connectionPayload },
      pingPayload = pingPayload,
      pongPayload = pongPayload,
      connectionAcknowledgeTimeoutMs = connectionAcknowledgeTimeoutMs,
      pingIntervalMillis = pingIntervalMillis,
      frameType = frameType,
      webSocketConnection = webSocketConnection,
      listener = listener,
      scope = scope,
  )

  override suspend fun connectionInit() {
    val message = mutableMapOf<String, Any?>(
        "type" to "connection_init",
    )

    val payload = connectionPayload()
    if (payload != null) {
      message.put("payload", payload)
    }

    sendMessageMap(message, frameType)

    withTimeout(connectionAcknowledgeTimeoutMs) {
      val map = receiveMessageMap()
      when (val type = map["type"]) {
        "connection_ack" -> return@withTimeout
        else -> println("unknown graphql-ws message while waiting for connection_ack: '$type")
      }
    }
  }

  override fun <D : Operation.Data> startOperation(request: ApolloRequest<D>) {
    sendMessageMap(
        mapOf(
            "type" to "subscribe",
            "id" to request.requestUuid.toString(),
            "payload" to DefaultHttpRequestComposer.composePayload(request)
        ),
        frameType
    )
  }

  override fun <D : Operation.Data> stopOperation(request: ApolloRequest<D>) {
    sendMessageMap(
        mapOf(
            "type" to "complete",
            "id" to request.requestUuid.toString()
        ),
        frameType
    )
  }

  override suspend fun run() {
    if (pingIntervalMillis > 0) {
      scope.launch {
        val map = mutableMapOf<String, Any?>(
            "type" to "ping",
        )
        if (pingPayload != null) {
          map["payload"] = pingPayload
        }

        while (true) {
          delay(pingIntervalMillis)
          sendMessageMap(map, frameType)
        }
      }
    }
    super.run()
  }

  override fun handleServerMessage(messageMap: Map<String, Any?>) {
    @Suppress("UNCHECKED_CAST")
    when (messageMap["type"]) {
      "next" -> listener.operationResponse(messageMap["id"] as String, messageMap["payload"] as Map<String, Any?>)
      "error" -> listener.operationError(messageMap["id"] as String, messageMap["payload"] as Map<String, Any?>)
      "complete" -> listener.operationComplete(messageMap["id"] as String)
      "ping" -> {
        val map = mutableMapOf<String, Any?>(
            "type" to "pong",
        )
        if (pongPayload != null) {
          map["payload"] = pongPayload
        }
        sendMessageMap(map, frameType)
      }
      "pong" -> Unit // Nothing to do, the server acknowledged one of our pings
      else -> Unit // Unknown message
    }
  }

  /**
   * A factory for [GraphQLWsProtocol].
   *
   * @param connectionPayload a map of additional parameters to send in the "connection_init" message
   * @param pingIntervalMillis the interval between two client-initiated pings or -1 to not send any ping.
   * Default value: -1
   * @param pingPayload the ping payload to send in "ping" messages or null to not send a payload
   * @param pongPayload the pong payload to send in "pong" messages or null to not send a payload
   * @param connectionAcknowledgeTimeoutMs the timeout for receiving the "connection_ack" message, in milliseconds
   * @param frameType the type of the websocket frames to use. Default value: [WsFrameType.Text]
   */
  class Factory(
      private val connectionPayload: suspend () -> Map<String, Any?>? = { null },
      private val pingIntervalMillis: Long = -1,
      private val pingPayload: Map<String, Any?>? = null,
      private val pongPayload: Map<String, Any?>? = null,
      private val connectionAcknowledgeTimeoutMs: Long = 10_000,
      private val frameType: WsFrameType = WsFrameType.Text,
  ) : WsProtocol.Factory {

    @Deprecated("Use the constructor with connectionPayload as a lambda instead",
        ReplaceWith("Factory({ connectionPayload }, pingIntervalMillis, pingPayload, pongPayload, connectionAcknowledgeTimeoutMs)"))
    @ApolloDeprecatedSince(v3_2_3)
    constructor(
        connectionPayload: Map<String, Any?>?,
        pingIntervalMillis: Long = -1,
        pingPayload: Map<String, Any?>? = null,
        pongPayload: Map<String, Any?>? = null,
        connectionAcknowledgeTimeoutMs: Long = 10_000,
        frameType: WsFrameType = WsFrameType.Text,
    ) : this(
        connectionPayload = { connectionPayload },
        pingIntervalMillis = pingIntervalMillis,
        pingPayload = pingPayload,
        pongPayload = pongPayload,
        connectionAcknowledgeTimeoutMs = connectionAcknowledgeTimeoutMs,
        frameType = frameType,
    )

    override val name: String
      get() = "graphql-transport-ws"

    override fun create(webSocketConnection: WebSocketConnection, listener: Listener, scope: CoroutineScope): WsProtocol {
      return GraphQLWsProtocol(
          connectionPayload = connectionPayload,
          pingPayload = pingPayload,
          pongPayload = pongPayload,
          connectionAcknowledgeTimeoutMs = connectionAcknowledgeTimeoutMs,
          pingIntervalMillis = pingIntervalMillis,
          frameType = frameType,
          webSocketConnection = webSocketConnection,
          listener = listener,
          scope = scope
      )
    }
  }
}
